[2.x] AssertionError: assertion failed: Not absolute

## steps

### build.sbt 

```scala
val foo = project in file("foo")
```

### foo/build.sbt 

```scala
val foo = project in file(".")
```

### project/build.properties 

```properties
sbt.version=2.0.0-RC12
```

## problem

```
java.lang.AssertionError: assertion failed: Not absolute: foo
	at scala.runtime.Scala3RunTime$.assertFailed(Scala3RunTime.scala:10)
	at sbt.io.IO$.assertAbsolute(IO.scala:1222)
	at sbt.io.IO$.resolve(IO.scala:1217)
	at sbt.internal.Load$.isRootPath(Load.scala:703)
	at sbt.internal.Load$.$anonfun$74(Load.scala:1378)
	at scala.collection.StrictOptimizedIterableOps.partition$$anonfun$1(StrictOptimizedIterableOps.scala:37)
	at scala.collection.IterableOnceOps.foreach(IterableOnce.scala:633)
	at scala.collection.IterableOnceOps.foreach$(IterableOnce.scala:337)
	at scala.collection.AbstractIterator.foreach(Iterator.scala:1328)
	at scala.collection.StrictOptimizedIterableOps.partition(StrictOptimizedIterableOps.scala:37)
	at scala.collection.StrictOptimizedIterableOps.partition$(StrictOptimizedIterableOps.scala:30)
	at scala.collection.immutable.List.partition(List.scala:582)
	at sbt.internal.Load$.discoverProjects(Load.scala:1378)
	at sbt.internal.Load$.discover$1(Load.scala:1068)
	at sbt.internal.Load$.discoverAndLoad$1(Load.scala:1111)
	at sbt.internal.Load$.loadTransitive(Load.scala:1128)
	at sbt.internal.Load$.loadTransitive1$1(Load.scala:1046)
	at sbt.internal.Load$.loadTransitive(Load.scala:1155)
	at sbt.internal.Load$.loadProjects$1(Load.scala:861)
	at sbt.internal.Load$.$anonfun$49(Load.scala:864)
	at sbt.internal.Load$.timed(Load.scala:1648)
	at sbt.internal.Load$.loadUnit$$anonfun$1(Load.scala:865)
	at sbt.internal.Load$.timed(Load.scala:1648)
	at sbt.internal.Load$.loadUnit(Load.scala:914)
	at sbt.internal.Load$.$anonfun$30$$anonfun$1(Load.scala:569)
	at sbt.internal.BuildLoader$.componentLoader$$anonfun$1$$anonfun$1$$anonfun$1$$anonfun$1(BuildLoader.scala:180)
	at sbt.internal.BuildLoader.apply(BuildLoader.scala:246)
	at sbt.internal.Load$.loadURI$1(Load.scala:636)
	at sbt.internal.Load$.loadAll(Load.scala:652)
	at sbt.internal.Load$.loadURI(Load.scala:582)
	at sbt.internal.Load$.load(Load.scala:555)
	at sbt.internal.Load$.$anonfun$8(Load.scala:269)
	at sbt.internal.Load$.timed(Load.scala:1648)
	at sbt.internal.Load$.apply(Load.scala:269)
	at sbt.internal.Load$.defaultLoad(Load.scala:60)
	at sbt.BuiltinCommands$.doLoadProject(Main.scala:979)
	at sbt.BuiltinCommands$.loadProjectImpl$$anonfun$2(Main.scala:933)
	at sbt.Command$.applyEffect$$anonfun$2$$anonfun$1(Command.scala:149)
	at sbt.Command$.applyEffect$$anonfun$1$$anonfun$1(Command.scala:144)
	at sbt.Command$.process(Command.scala:188)
	at sbt.MainLoop$.$anonfun$14(MainLoop.scala:269)
	at scala.Option.getOrElse(Option.scala:203)
	at sbt.MainLoop$.process$1(MainLoop.scala:269)
	at sbt.MainLoop$.processCommand(MainLoop.scala:326)
	at sbt.MainLoop$.next$$anonfun$1$$anonfun$1(MainLoop.scala:172)
	at sbt.State$StateOpsImpl$.runCmd$1(State.scala:270)
	at sbt.State$StateOpsImpl$.process$extension(State.scala:306)
	at sbt.MainLoop$.next$$anonfun$1(MainLoop.scala:172)
	at sbt.internal.util.ErrorHandling$.wideConvert(ErrorHandling.scala:24)
	at sbt.MainLoop$.next(MainLoop.scala:173)
	at sbt.MainLoop$.runLoop(MainLoop.scala:148)
	at sbt.MainLoop$.run(MainLoop.scala:143)
	at sbt.MainLoop$.runWithNewLog$$anonfun$1(MainLoop.scala:121)
	at sbt.io.Using.apply(Using.scala:41)
	at sbt.MainLoop$.runWithNewLog(MainLoop.scala:114)
	at sbt.MainLoop$.runAndClearLast(MainLoop.scala:70)
	at sbt.MainLoop$.runLoggedLoop(MainLoop.scala:53)
	at sbt.MainLoop$.runLogged(MainLoop.scala:46)
	at sbt.StandardMain$.runManaged(Main.scala:233)
	at sbt.xMain$.run$$anonfun$3(Main.scala:143)
	at scala.util.DynamicVariable.withValue(DynamicVariable.scala:60)
	at scala.Console$.withIn(Console.scala:229)
	at scala.Console$.withIn(Console.scala:242)
	at sbt.internal.util.Terminal$.withIn(Terminal.scala:642)
	at sbt.internal.util.Terminal$.withStreams$$anonfun$1(Terminal.scala:420)
	at scala.util.DynamicVariable.withValue(DynamicVariable.scala:60)
	at scala.Console$.withOut(Console.scala:166)
	at sbt.internal.util.Terminal$.withOut$$anonfun$2(Terminal.scala:632)
	at scala.util.DynamicVariable.withValue(DynamicVariable.scala:60)
	at scala.Console$.withErr(Console.scala:195)
	at sbt.internal.util.Terminal$.withOut(Terminal.scala:632)
	at sbt.internal.util.Terminal$.withStreams(Terminal.scala:420)
	at sbt.xMain$.withStreams$1(Main.scala:94)
	at sbt.xMain$.run(Main.scala:145)
	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103)
	at java.base/java.lang.reflect.Method.invoke(Method.java:580)
	at sbt.internal.XMainConfiguration.run(XMainConfiguration.java:68)
	at sbt.xMain.run(Main.scala:48)
	at xsbt.boot.Launch$.run$$anonfun$1(Launch.scala:132)
	at xsbt.boot.Launch$.withContextLoader(Launch.scala:157)
	at xsbt.boot.Launch$.run(Launch.scala:132)
	at xsbt.boot.Launch$.apply$$anonfun$1(Launch.scala:43)
	at xsbt.boot.Launch$.launch(Launch.scala:142)
	at xsbt.boot.Launch$.apply(Launch.scala:43)
	at xsbt.boot.Launch$.apply(Launch.scala:24)
	at xsbt.boot.Boot$.runImpl(Boot.scala:73)
	at xsbt.boot.Boot$.run(Boot.scala:69)
	at xsbt.boot.Boot$.main(Boot.scala:23)
	at xsbt.boot.Boot.main(Boot.scala)
[error] java.lang.AssertionError: assertion failed: Not absolute: foo
[error] Use 'last' for the full log.
```

## expectation

<!-- Describe what you think should've happened. -->

## notes

- https://github.com/sbt/sbt/blob/0968cbf7eff88b65dd2e68ca23b3b96cd78c4636/build.sbt#L1426
- https://github.com/sbt/sbt/blob/0968cbf7eff88b65dd2e68ca23b3b96cd78c4636/launcher-package/build.sbt#L96
- https://github.com/sbt/sbt/issues/9010

--- a/buildfile/src/main/scala/sbt/internal/EvaluateConfigurations.scala
+++ b/buildfile/src/main/scala/sbt/internal/EvaluateConfigurations.scala
@@ -172,8 +172,7 @@ private[sbt] object EvaluateConfigurations {
         val compositeProjects = definitions
           .values(loader)
           .collect { case p: CompositeProject => p }
-        // todo: resolveBase?
-        CompositeProject.expand(compositeProjects) // .map(resolveBase(file.getParentFile, _))
+        CompositeProject.expand(compositeProjects)
       }
       val loadedDslEntries = dslEntries.map(_.result.apply(loader))
       val settings = loadedDslEntries.collect { case DslEntry.ProjectSettings(s) => s }.flatten

--- a/buildfile/src/main/scala/sbt/internal/LoadedSbtFile.scala
+++ b/buildfile/src/main/scala/sbt/internal/LoadedSbtFile.scala
@@ -16,15 +16,15 @@ import java.nio.file.Path
  * Represents the exported contents of a .sbt file.  Currently, that includes the list of settings,
  * the values of Project vals, and the import statements for all defined vals/defs.
  */
-private[sbt] final class LoadedSbtFile(
-    val settings: Seq[Setting[?]],
-    val projects: Seq[Project],
-    val importedDefs: Seq[String],
-    val manipulations: Seq[Project => Project],
+private[sbt] final case class LoadedSbtFile(
+    settings: Seq[Setting[?]],
+    projects: Seq[Project],
+    importedDefs: Seq[String],
+    manipulations: Seq[Project => Project],
     // TODO - we may want to expose a simpler interface on top of here for the set command,
     // rather than what we have now...
-    val definitions: DefinedSbtValues,
-    val generatedFiles: Seq[Path]
+    definitions: DefinedSbtValues,
+    generatedFiles: Seq[Path]
 ) {
   // We still use merge for now.  We track originating sbt file in an alternative manner.
   def merge(o: LoadedSbtFile): LoadedSbtFile =

--- a/main/src/main/scala/sbt/internal/Load.scala
+++ b/main/src/main/scala/sbt/internal/Load.scala
@@ -31,7 +31,7 @@ import sbt.io.{ GlobFilter, IO }
 import sbt.librarymanagement.{ Configuration, Configurations, Resolver, ScalaArtifacts }
 import sbt.nio.Settings
 import sbt.util.{ Logger, Show }
-import xsbti.{ FileConverter, HashedVirtualFileRef, VirtualFile }
+import xsbti.{ FileConverter, HashedVirtualFileRef, PathBasedFile, VirtualFile }
 import xsbti.compile.{ ClasspathOptionsUtil, Compilers }
 import java.io.File
 import java.net.URI
@@ -1340,13 +1340,19 @@ private[sbt] object Load {
     // TODO - We should import vals defined in other sbt files here, if we wish to
     // share.  For now, build.sbt files have their own unique namespace.
     def loadSettingsFile(src: VirtualFile): LoadedSbtFile =
-      EvaluateConfigurations.evaluateSbtFile(
+      val evaluated = EvaluateConfigurations.evaluateSbtFile(
         eval(),
         src,
         IO.readStream(src.input()).linesIterator.toList,
         loadedPlugins.detected.imports,
         0
       )(loader)
+      evaluated.copy(
+        projects = evaluated.projects.map: p =>
+          src match
+            case file: PathBasedFile => resolveBase(file.toPath.toFile.getParentFile)(p)
+            case _                   => p
+      )
     // How to merge SbtFiles we read into one thing
     def merge(ls: Seq[LoadedSbtFile]): LoadedSbtFile = ls.foldLeft(LoadedSbtFile.empty) {
       _.merge(_)

--- a/sbt-app/src/sbt-test/project-load/subdirectory/build.sbt
+++ b/sbt-app/src/sbt-test/project-load/subdirectory/build.sbt
@@ -1,0 +1,10 @@
+lazy val foo = project
+
+@transient
+lazy val check = taskKey[Unit]("")
+
+scalaVersion := "3.8.3"
+LocalRootProject / check := {
+  assert((foo / Compile / scalacOptions).value == List("-Xmacro-settings:a:a"),
+    s"${(foo / Compile / scalacOptions).value}")
+}

--- a/sbt-app/src/sbt-test/project-load/subdirectory/changes/Good.scala
+++ b/sbt-app/src/sbt-test/project-load/subdirectory/changes/Good.scala
@@ -1,0 +1,3 @@
+package example
+
+class A

--- a/sbt-app/src/sbt-test/project-load/subdirectory/foo/A.scala
+++ b/sbt-app/src/sbt-test/project-load/subdirectory/foo/A.scala
@@ -1,0 +1,3 @@
+package example
+
+class A("

--- a/sbt-app/src/sbt-test/project-load/subdirectory/foo/build.sbt
+++ b/sbt-app/src/sbt-test/project-load/subdirectory/foo/build.sbt
@@ -1,0 +1,4 @@
+lazy val foo = rootProject
+  .settings(
+    Compile / scalacOptions += "-Xmacro-settings:a:a"
+  )

--- a/sbt-app/src/sbt-test/project-load/subdirectory/test
+++ b/sbt-app/src/sbt-test/project-load/subdirectory/test
@@ -1,0 +1,6 @@
+> check
+
+# check that it fails to compile a bad source
+-> compile
+$ copy-file changes/Good.scala foo/A.scala
+> compile

